src/core/models/outputs/output_foundry.c

```diff
@@ -1,4 +1,413 @@
+      return NULL;
+    }
+    offset += native_tokens_serialize_len(&output->native_tokens);
+  } else {
+    offset += sizeof(uint16_t);
+  }
+
+  // serial number
+  if (buf_len < offset + sizeof(uint32_t)) {
+    printf("[%s:%d] invalid data length\n", __func__, __LINE__);
+    output_foundry_free(output);
+    return NULL;
+  }
+  memcpy(&output->serial, &buf[offset], sizeof(uint32_t));
+  offset += sizeof(uint32_t);
+
+  // token tag
+  if (buf_len < offset + TOKEN_TAG_BYTES_LEN) {
+    printf("[%s:%d] invalid data length\n", __func__, __LINE__);
+    output_foundry_free(output);
+    return NULL;
+  }
+  memcpy(&output->token_tag, &buf[offset], TOKEN_TAG_BYTES_LEN);
+  offset += TOKEN_TAG_BYTES_LEN;
+
+  // circulating supply
+  if (buf_len < offset + sizeof(uint256_t)) {
+    printf("[%s:%d] invalid data length\n", __func__, __LINE__);
+    output_foundry_free(output);
+    return NULL;
+  }
+  memcpy(&output->circ_supply, &buf[offset], sizeof(uint256_t));
+  offset += sizeof(uint256_t);
+
+  // maximum supply
+  if (buf_len < offset + sizeof(uint256_t)) {
+    printf("[%s:%d] invalid data length\n", __func__, __LINE__);
+    output_foundry_free(output);
+    return NULL;
+  }
+  memcpy(&output->max_supply, &buf[offset], sizeof(uint256_t));
+  offset += sizeof(uint256_t);
+
+  // token scheme
+  if (buf_len < offset + sizeof(uint8_t)) {
+    printf("[%s:%d] invalid data length\n", __func__, __LINE__);
+    output_foundry_free(output);
+    return NULL;
+  }
+  memcpy(&output->token_scheme, &buf[offset], sizeof(uint8_t));
+  offset += sizeof(uint8_t);
+
+  // feature blocks
+  uint8_t feat_block_count = *((uint8_t*)&buf[offset]);
+  if (feat_block_count > 0) {
+    output->feature_blocks = feat_blk_list_deserialize(&buf[offset], buf_len - offset);
+    if (!output->feature_blocks) {
+      printf("[%s:%d] can not deserialize feature blocks\n", __func__, __LINE__);
+      output_foundry_free(output);
+      return NULL;
+    }
+  }
+
+  return output;
+}
+
+void output_foundry_print(output_foundry_t* output) {
+  if (output == NULL) {
+    printf("[%s:%d] invalid parameters\n", __func__, __LINE__);
+    return;
+  }
+
+  printf("Foundry Output: [\n");
+
+  printf("\tAddress: ");
+  address_print(output->address);
+
+  printf("\tAmount: %" PRIu64 "\n", output->amount);
+
+  // print native tokens
+  native_tokens_t *token, *tmp;
+  char* amount_str;
+  printf("\tNative Tokens: [\n");
+  HASH_ITER(hh, *(&output->native_tokens), token, tmp) {
+    amount_str = uint256_to_str(token->amount);
+    printf("\t\t[%s] ", amount_str);
+    dump_hex_str(token->token_id, NATIVE_TOKEN_ID_BYTES);
+    free(amount_str);
+  }
+  printf("\t]\n");
+
+  printf("\tSerial Number: %" PRIu32 "\n", output->serial);
+
+  // print token tag
+  printf("\tToken Tag: ");
+  dump_hex_str(output->token_tag, TOKEN_TAG_BYTES_LEN);
+
+  // print circulating supply
+  char* circ_supply_str;
+  circ_supply_str = uint256_to_str(&output->circ_supply);
```

Comment:
A check for a NULL pointer should be made.

Reply:
Done.